da/da.go

```diff
@@ -19,22 +20,41 @@ const (
 	StatusError
 )
 
-// ResultSubmitBlock contains information returned from DA layer after block submission.
-type ResultSubmitBlock struct {
+type DAResult struct {
 	// Code is to determine if the action succeeded.
 	Code StatusCode
-	// Message may contain DA layer specific information (like detailed error message)
+	// Message may contain DA layer specific information (like DA block height/hash, detailed error message, etc)
 	Message string
+}
+
+// ResultSubmitBlock contains information returned from DA layer after block submission.
+type ResultSubmitBlock struct {
+	DAResult
 	// Not sure if this needs to be bubbled up to other
 	// parts of Optimint.
 	// Hash hash.Hash
 }
 
+// ResultCheckBlock contains information about block availability, returned from DA layer client.
+type ResultCheckBlock struct {
+	DAResult
+	// DataAvailable is the actual answer whether the block is available or not.
+	// It can be true if and only if Code is equal to StatusSuccess.
+	DataAvailable bool
+}
+
+type ResultRetrieveBlock struct {
+	DAResult
+	// Block is the full block retrieved from Data Availability Layer.
+	// If Code is not equal to StatusSuccess, it has to be nil.
+	Block *types.Block
+}
+
 // DataAvailabilityLayerClient defines generic interface for DA layer block submission.
 // It also contains life-cycle methods.
 type DataAvailabilityLayerClient interface {
 	// Init is called once to allow DA client to read configuration and initialize resources.
-	Init(config []byte, logger log.Logger) error
+	Init(config []byte, kvStore store.KVStore, logger log.Logger) error
```

Comment:
Question: What's the reason this 'contructor-like' method is part of the interface? Can this be just an implementation specific constructor?

Reply:
My goal is to be able to store DA layer in configuration (file). With such an generic constructor, any DALC can be initialized in uniform way. So after implementing new DALC, you only need to change configuration, without any code modification in optimint.